crates/spirv-std/macros/src/lib.rs

```diff
@@ -625,3 +625,132 @@ fn debug_printf_inner(input: DebugPrintfInput) -> TokenStream {
+                t.push(TokenTree::Literal(proc_macro2::Literal::string(s.as_str())));
+                t.push(TokenTree::Punct(proc_macro2::Punct::new(
+                    ',',
+                    proc_macro2::Spacing::Alone,
+                )))
+            }
+        }
+    }
+
+    fn add_regs(&self, t: &mut Vec<TokenTree>) {
+        for i in 0..SAMPLE_PARAM_COUNT {
+            if self.0 & (1 << i) != 0 {
+                let s = format!("{0} = in(reg) &param.{0}", SAMPLE_PARAM_NAMES[i]);
+                t.push(TokenTree::Literal(proc_macro2::Literal::string(s.as_str())));
+                t.push(TokenTree::Punct(proc_macro2::Punct::new(
+                    ',',
+                    proc_macro2::Spacing::Alone,
+                )))
+            }
+        }
+    }
+}
+
+impl syn::visit_mut::VisitMut for SampleFnRewriter {
+    fn visit_macro_mut(&mut self, m: &mut syn::Macro) {
+        if m.path.is_ident("asm") {
+            let t = m.tokens.clone();
+            let mut new_t = Vec::new();
+            let mut altered = false;
+
+            for tt in t {
+                match tt {
+                    TokenTree::Literal(l) => {
+                        if let Ok(l) = syn::parse::<syn::LitStr>(l.to_token_stream().into()) {
+                            let s = l.value();
+                            if s.contains("$PARAMS") {
+                                altered = true;
+                                self.add_loads(&mut new_t);
+                                let s = s.replace("$PARAMS", &self.get_operands());
+                                new_t.push(TokenTree::Literal(proc_macro2::Literal::string(
+                                    s.as_str(),
+                                )));
+                            } else {
+                                new_t.push(TokenTree::Literal(l.token()));
+                            }
+                        } else {
+                            new_t.push(TokenTree::Literal(l));
+                        }
+                    }
+                    _ => {
+                        new_t.push(tt);
+                    }
+                }
+            }
+
+            if altered {
+                self.add_regs(&mut new_t);
+            }
+
+            m.tokens = new_t.into_iter().collect();
+        }
+    }
+}
+
+/// Generates permutations of a sampling function containing inline assembly with an image
+/// instruction ending with a placeholder `$PARAMS` operand. The last parameter must be named
+/// `params`, its type will be rewritten. Relevant generic arguments are added to the function
+/// signature. See `SAMPLE_PARAM_TYPES` for a list of names you cannot use as generic arguments.
+#[proc_macro_attribute]
+#[doc(hidden)]
+pub fn gen_sample_param_permutations(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item_fn = syn::parse_macro_input!(item as syn::ItemFn);
+    let mut fns = Vec::new();
+
+    for m in 1..((1 << SAMPLE_PARAM_COUNT) - 1) {
```

Comment:
```suggestion
    for m in 1..(1 << SAMPLE_PARAM_COUNT) {
```
(or `..=` with the `- 1`)

Reply:
Oh good call. I had it correct when I was still hardcoding those numbers 😄